### PR TITLE
[Ray Usage Stats] Record cluster metadata + Refactoring.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This is the first PR to implement usage stats on Ray. Please refer to the file `usage_lib.py` for more details.

The full specification is here https://docs.google.com/document/d/1ZT-l9YbGHh-iWRUC91jS-ssQ5Qe2UQ43Lsoc1edCalc/edit#heading=h.17dss3b9evbj.

You can see the full PR for phase 1 from here; https://github.com/rkooo567/ray/pull/108/files.

The PR is doing some basic refactoring + adding cluster metadata to GCS instead of the version numbers. 

After this PR, we will add code to enable usage report "off by default". 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/usage/usage_lib.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+"Ray API server (currently a dashboard server)" reports the usage data to the hard-coded URL.
+`ray._private.usage.usage_constants.USAGE_REPORT_URL`.
+
+Data is reported
+- Every `ray._private.usage.usage_constants.USAGE_REPORT_INTERVAL_S`.
+
+Note that it is also possible to configure the interval using the environment variable,
+`RAY_USAGE_REPORT_INTERVAL_S`.
+
+To see collected/reported data, see `ray._private.usage.schema.py`.
```
Comment (jjyao) [inline python/ray/_private/usage/usage_lib.py:43]: see usage_stats.json

**Review by jjyao — APPROVED** [review]

**Comment by rkooo567** [comment]
cc @richardliaw

File: python/ray/_private/utils.py [hunk 2]
```diff
@@ -1197,6 +1197,58 @@ def check_dashboard_dependencies_installed() -> bool:
...
+def internal_kv_put_with_retry(gcs_client, key, value, namespace, num_retries=20):
+    if isinstance(key, str):
+        key = key.encode()
+    error = None
+    for _ in range(num_retries):
+        try:
+            return gcs_client.internal_kv_put(
+                key, value, overwrite=True, namespace=namespace
+            )
+        except grpc.RpcError as e:
```
Comment (simon-mo) [inline python/ray/_private/utils.py:1233]: why are we catching RpcError here vs Exception in internal_kv_get_with_retry

File: python/requirements/requirements_default.txt [hunk 3]
```diff
@@ -1,9 +1,7 @@
 aiohttp>=3.7
 aiosignal
 aiohttp_cors
-aioredis<2
 colorful
-frozenlist
```
Comment (simon-mo) [inline python/requirements/requirements_default.txt:6]: intentional?

File: python/ray/_private/usage/usage_lib.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+# Internal APIs #
+#################
+
+
+def _generate_cluster_metadata():
+    """Return a dictionary of cluster metadata."""
+    ray_version, python_version = ray._private.utils.compute_version_info()
+    return {
+        "schema_version": usage_constant.SCHEMA_VERSION,
+        "source": os.getenv("RAY_USAGE_STATS_SOURCE", "OSS"),
```
Comment (simon-mo) [inline python/ray/_private/usage/usage_lib.py:70]: confirm with @richardliaw this is the source value we want by default?

**Review by simon-mo — APPROVED** [review]
Overall looks good. One question, I can't find why are we piping `session_dir` throughout this PR? I probably missed it but can't find a place it's actually used by usage_stage.

File: python/requirements/requirements_default.txt (hunk 3 above)
Comment (rkooo567) [reply python/requirements/requirements_default.txt:6]: This had to be removed long time ago.

File: python/ray/_private/utils.py (hunk 2 above)
Comment (rkooo567) [reply python/ray/_private/utils.py:1233]: Hmm I am not sure about the reasoning actually. This part is just refactoring. I am planning to move this util method into gcs_client and add num_retires instead, so we can talk about more details in that PR.

File: python/ray/_private/usage/usage_lib.py (hunk 4 above)
Comment (rkooo567) [reply python/ray/_private/usage/usage_lib.py:70]: I will merge the PR first. @richardliaw lmk what you think!

**Pull request merged**